reqwest-tracing/src/reqwest_otel_span_builder.rs

```diff
@@ -0,0 +1,113 @@
+use reqwest::{
+    header::{HeaderMap, HeaderValue},
+    Request, Response, StatusCode as RequestStatusCode,
+};
+use reqwest_middleware::{Error, Result};
+use task_local_extensions::Extensions;
+use tracing::Span;
+
+use crate::{impl_on_request_failure, impl_on_request_success, reqwest_otel_span};
+
+/// [`RequestOtelSpanBackend`] allows you to customise the span attached by
+/// [`TracingMiddleware`] to incoming requests.
+///
+/// Check out [`reqwest_otel_span`] documentation for examples.
+///
+/// [`TracingMiddleware`]: crate::middleware::TracingMiddleware.
+pub trait RequestOtelSpanBackend {
+    /// Initalized a new span before the request is executed.
+    fn on_request_start(req: &Request, extension: &mut Extensions) -> Span;
+
+    /// Runs after the request call has executed.
+    fn on_request_end(span: &Span, outcome: &Result<Response>, extension: &mut Extensions) {}
+
+    /// Runs only on a successful request.
+    fn on_request_success(span: &Span, response: &Response, extension: &mut Extensions);
+
+    /// Runs only on a failed request.
+    fn on_request_failure(span: &Span, e: &Error, _extension: &mut Extensions);
```

Comment:
I am unconvinced - this can be done in a single handler that matches on the outcome.
If you want to override the behaviour just for successes, you can do:

```
fn on_request_end(...) {
   if let Err(error) = outcome {
       // custom logic
   } else {
       // invoke default handler
   }
}
```